astropy/units/quantity.py

```diff
@@ -1363,14 +1363,12 @@ def __int__(self):
     def __index__(self):
         # for indices, we do not want to mess around with scaling at all,
         # so unlike for float, int, we insist here on unscaled dimensionless
-        try:
-            assert self.unit.is_unity()
-            return self.value.__index__()
-        except Exception:
+        if not (self.unit.is_unity() and hasattr(self.value, "__index__")):
```

Comment:
personally not a fan of walrus operators buried too far inside expressions

is

```py
as_index = getattr(self.value, "__index__", None)
if as_index is None or not self.unit.is_unity():
    raise TypeError(...)

return as_index()
```

ok?

Reply:
Definitely fine, yes!

Reply:
Continuing a theme of minimizing the logic changes in this PR, why not *ONLY* change the `assert` statement? Replace `assert self.unit.is_unity()` with:
```
if not self.unit.is_unity():
    raise AssertionError
```
This will retain the original logic with no ambiguity. If the original code is wrong somehow then it would seem better to address that in a separate narrow PR.